core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -230,7 +242,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
         val reply = SparkAppConfig(
           sparkProperties,
           SparkEnv.get.securityManager.getIOEncryptionKey(),
-          fetchHadoopDelegationTokens())
+          Option(delegationTokens.get()))
```

Comment:
realized the other changes were small so I took a look -- this is the only part which concerns me.  This used to be a blocking call, so you were sure that you'd get tokens to respond to `RetrieveSparkAppConfig`.  However, now you're getting the delegation tokens async.  You will probably get them back from the TokenManager before sending them here, but there is a race, right?  Will it be OK if the executor doesn't get the tokens here?  I know you'll send the executor the tokens in an `UpdateDelegationTokens` msg later, but will that lead to some spurious failures?

you could change the token manager to call `driver.ask(UpdateDelegationTokens(tokens))` inside `start()`, and ensure that future is completed here, though it makes things more complex.

Reply:
Although it's not obvious I think this is fine.

The token manager is started before any rm-specific code runs and requests executors (in the `CoarseGrainedSchedulerBackend.start()` method; rm-specific schedulers call that before they do anything). So the message to set these tokens is the first message queued on the driver, and this reference will already be set when the first executor asks for this info.

Reply:
Actually since I moved the token manager to `CoarseGrainedSchedulerBackend` I think I can simplify this and avoid having to trace how the RPC code works to prove this is fine. I'll try to update things this afternoon after I finish some other stuff.

Reply:
Not sure I follow.  to be clear, I definitely think this is extremely unlikely, but I don't see what totally prevents it.

>  So the message to set these tokens is the first message queued on the driver, and this reference will already be set when the first executor asks for this info.

I see that the message will get queued from the TokenManager to get sent to the SchedulerBackend before anything else happens.  But I was a bit hazy on what happens after that -- I guess because both are local, the way the rpc env works, nothing else can squeeze in-between?  its definitely not obvious -- seems its worth at least a short comment like "Because the  token manager is local, we know we will have already received the initial set of tokens (if any)"

Reply:
>  But I was a bit hazy on what happens after that

The RPC layer is basically a message queue. So first message to be sent = first message to be processed.

But as I said, I'll clean this up since now the code is all in `CoarseGrainedSchedulerBackend` and that makes it easier.